From 524cddbaa986b53106931ed62c89ea31dbc6afa6 Mon Sep 17 00:00:00 2001
From: Eneas U de Queiroz <cote2004-github@yahoo.com>
Date: Thu, 25 Oct 2018 09:17:38 -0300
Subject: [PATCH] eng_devcrypto: disable digests support

Digest support is broken.
In digest_cleanup(), EVP_CIPHER_CTX_md_data(ctx) returns NULL, and then
is dereferenced.

Digests may be reenabled by defining USE_DEVCRYPTO_DIGESTS.

Signed-off-by: Eneas U de Queiroz <cote2004-github@yahoo.com>

diff --git a/crypto/engine/eng_devcrypto.c b/crypto/engine/eng_devcrypto.c
index 51105ec60d..25a9d3e032 100644
--- a/crypto/engine/eng_devcrypto.c
+++ b/crypto/engine/eng_devcrypto.c
@@ -333,7 +333,7 @@ static int devcrypto_ciphers(ENGINE *e, const EVP_CIPHER **cipher,
  * perilous if there's a lot of data coming in (if someone wants to checksum
  * an OpenSSL tarball, for example).
  */
-#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL)
+#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL) && defined(USE_DEVCRYPTO_DIGESTS)
 
 /******************************************************************************
  *
@@ -598,7 +598,7 @@ static int devcrypto_digests(ENGINE *e, const EVP_MD **digest,
 static int devcrypto_unload(ENGINE *e)
 {
     destroy_all_cipher_methods();
-#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL)
+#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL) && defined(USE_DEVCRYPTO_DIGESTS)
     destroy_all_digest_methods();
 #endif
     return 1;
@@ -618,7 +618,7 @@ void engine_load_devcrypto_int()
     }
 
     prepare_cipher_methods();
-#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL)
+#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL) && defined(USE_DEVCRYPTO_DIGESTS)
     prepare_digest_methods();
 #endif
 
@@ -664,7 +664,7 @@ void engine_load_devcrypto_int()
 # endif
 #endif
         || !ENGINE_set_ciphers(e, devcrypto_ciphers)
-#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL)
+#if defined(COP_FLAG_UPDATE) && defined(COP_FLAG_FINAL) && defined(USE_DEVCRYPTO_DIGESTS)
         || !ENGINE_set_digests(e, devcrypto_digests)
 #endif
         ) {
